lib/consul/async/json_endpoint.rb

```diff
@@ -0,0 +1,216 @@
+        _enable_network_debug if conf.debug && conf.debug[:network]
+        fetch
+        queue << Object.new
+      end
+
+      def _enable_network_debug
+        on_response do |result|
+          stats = result.stats
+          STDERR.puts "[DBUG][ OK ]#{result.modified? ? '[MODFIED]' : '[NO DIFF]'}" \
+          "[s:#{stats.successes},err:#{stats.errors}]" \
+          "[#{stats.body_bytes_human.ljust(8)}][#{stats.bytes_per_sec_human.ljust(9)}]"\
+          " #{url.ljust(48)}, next in #{result.retry_in} s"
+        end
+        on_error { |http| STDERR.puts "[ERROR]: #{url}: #{http.error.inspect}" }
+      end
+
+      def on_response(&block)
+        @s_callbacks << block
+      end
+
+      def on_error(&block)
+        @e_callbacks << block
+      end
+
+      def ready?
+        @ready
+      end
+
+      def terminate
+        @stopping = true
+      end
+
+      private
+
+      def build_request
+        res = {
+          head: {
+            'Accept'         => 'application/json',
+          },
+          url: url,
+          keepalive: true,
+          callback: method(:on_response)
+        }
+        res[:head]['accept-encoding'] = 'identity' unless conf.enable_gzip_compression
+        @query_params.each_pair do |k, v|
+          res[:query][k] = v
+        end
+        res
+      end
+
+      def _compute_retry_in(retry_in)
+        retry_in / 2 + Consul::Async::Utilities.random.rand(retry_in)
+      end
+
+      def _handle_error(http)
+        retry_in = _compute_retry_in([600, conf.retry_duration + 2**@consecutive_errors].min)
+        ::Consul::Async::Debug.puts_error "[#{url}] - #{http.error} - Retry in #{retry_in}s #{stats.body_bytes_human}"
+        @consecutive_errors += 1
+        http_result = HttpResponse.new(http)
+        EventMachine.add_timer(retry_in) do
+          yield
+          queue.push(Object.new)
+        end
+        @e_callbacks.each { |c| c.call(http_result) }
+      end
+
+      def fetch
+        options = {
+          connect_timeout: 5, # default connection setup timeout
+          inactivity_timeout: 60, # default connection inactivity (post-setup) timeout
+        }
+        connection = {
+          conn: EventMachine::HttpRequest.new(conf.url, options)
+        }
+        cb = proc do
+          http = connection[:conn].get(build_request)
+          http.callback do
+            if enforce_json_200 && http.response_header.status != 200 && http.response_header['Content-Type'] != 'application/json'
+              _handle_error(http) do
+                STDERR.puts "[RETRY][#{url}] (#{@consecutive_errors} errors)" if (@consecutive_errors % 10) == 1
+              end
+            else
+              @consecutive_errors = 0
+              http_result = HttpResponse.new(http)
+              new_content = http_result.response.freeze
+              modified = @last_result.fake? || @last_result.data != new_content
+              retry_in = modified ? conf.min_duration : conf.retry_on_non_diff
```

Comment:
Those only make sense for pre-defined consul endpoints, you might however add an optional value (intialized for instance with 60s of default retry time, so by default no consecutive queries more often than 60s, except if user specifies an update interval different.

Reply:
Actually having a different retry value based on whether last fetch revealed a new value might make sense (even for json endpoints).

If you had a change in the last fetch => retry quickly (just in case it still moves). Otherwise, assume it might not change soon enough.

Reply:
I've kept configuration in the inner class but do not expose it to user for now.